### PR TITLE
New creature abilities + magic blocking bonus

### PR DESCRIPTION
**Pull request by dydzio0614**
Does not compile yet - CBattleInfoCallback::battleGetAttackedHexes - cannot call "getPotentiallyAttackableHexes" with 3 parameters.

Includes feature requests from "new alternatives" mod authors

### CODE REVIEW

**Comment by Alucard648** [comment]
What about ability like "destroy 50% of target creature stack, regardless of actual damage"?

File: lib/battle/CBattleInfoCallback.cpp [hunk 1]
```diff
@@ -1647,6 +1647,24 @@ int CBattleInfoCallback::battleGetSurrenderCost(PlayerColor Player) const
...
+		node = getBattleNode();
+
+	if(!node)
+		return 1;
+
+	auto b = node->getBonuses(Selector::type(Bonus::BLOCK_MAGIC_BELOW));
+	if(b->size())
+		return b->totalValue();
+
+	return 1;
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: this should be `0`

**Comment by dydzio0614** [comment]
This branch is feature complete. Abilities not tested yet, so there may be bugs. The issues this branch has:
1. FIRST_STRIKE ability implementation is poor copypaste.
2. SYNERGY_TARGET is blank useless bonus to satisfy the needs of one modding team.
3. TERMINATOR is kind of reimplementation of DEATH_STARE, except that it uses different effect and works on undeads, golems etc.

**Comment by ArseniyShestakov** [comment]
I think we should require to provide testing maps / mods for PR like this one.
Also please add information about new bonuses into changelog.

**Comment by dydzio0614** [comment]
Map to test destruction ability: https://www.dropbox.com/s/lx4xyoocl8sm4vs/VCMI_DestCreAbTest.h3m?dl=0

Replacement for castle.json to test destruction: https://gist.github.com/dydzio0614/f956fe425811f21171472e5373912902 (desired effect - pikeman: 100% chance to kill 25 units after attack, halberdier: 50% chance to kill 25% of units after attack, archer: 100% chance to kill 25% of units)

Replacement for castle.json to test some other abilities: https://gist.github.com/dydzio0614/f789795c8fd2e14b18b8f84a83780d53 (pikeman: first strike, halberdier: wide breath, archer: attacks all adjacent)

**Comment by DjWarmonger** [comment]
Will check in free time.

File: lib/battle/CBattleInfoCallback.cpp [hunk 2]
```diff
@@ -186,7 +186,7 @@ std::set<BattleHex> CBattleInfoCallback::battleGetAttackedHexes(const CStack* at
 {
 	std::set<BattleHex> attackedHexes;
 	RETURN_IF_NOT_BATTLE(attackedHexes);
-
+
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: Excess whitespace

File: lib/battle/CBattleInfoCallback.cpp [hunk 3]
```diff
@@ -1157,44 +1157,85 @@ AttackableTiles CBattleInfoCallback::getPotentiallyAttackableHexes (const CStack
...
+		}
+		for(BattleHex tile : hexes)
+		{
+			//friendly stacks can also be damaged by Dragon Breath
+			if(battleGetStackByPos(tile, true))
+				if(battleGetStackByPos(tile, true) != attacker)
+					at.friendlyCreaturePositions.insert(tile);
+		}
+	}
+
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: Remove this newline

File: lib/battle/CBattleInfoCallback.cpp [hunk 4]
```diff
@@ -1157,44 +1157,85 @@ AttackableTiles CBattleInfoCallback::getPotentiallyAttackableHexes (const CStack
...
 		}
 	}
-	if (attacker->hasBonusOfType(Bonus::TWO_HEX_ATTACK_BREATH) && BattleHex::mutualPosition (destinationTile.hex, hex) > -1) //only adjacent hexes are subject of dragon breath calculation
+
+	if(attacker->hasBonusOfType(Bonus::WIDE_BREATH))
+	{
+		std::vector<BattleHex> hexes = destinationTile.neighbouringTiles();
+		for(int i = 0; i<hexes.size(); i++)
+		{
+			if(hexes.at(i) == hex) {
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: Put `{` on new line

File: lib/battle/CBattleInfoCallback.cpp [hunk 5]
```diff
@@ -1606,6 +1647,24 @@ int CBattleInfoCallback::battleGetSurrenderCost(PlayerColor Player) const
 	return ret;
 }
 
+si8 CBattleInfoCallback::battleMinSpellLevel(ui8 side) const
+{
+	const IBonusBearer *node = nullptr;
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: Pus space around `*`

**Pull request merged**